File: core/core-domain/src/main/java/com/coffee/gu/coupon/IssuedCoupon.java
package com.coffee.gu.coupon;

import com.coffee.gu.Principal;
import com.coffee.gu.enums.IssuedCouponState;

public class IssuedCoupon {
    private Long id;
    private Principal principal;
    private IssuedCouponState state;
    private Coupon coupon;

    public IssuedCoupon(Long id, Principal principal, IssuedCouponState state, Coupon coupon) {
        this.id = id;
        this.principal = principal;
        this.state = state;
        this.coupon = coupon;
    }

    public static IssuedCoupon download(Principal principal, Coupon coupon) {
        return new IssuedCoupon(null, principal, IssuedCouponState.DOWNLOADED, coupon);
    }

    public Long getId() {
        return id;
    }

    public IssuedCouponState getState() {
        return state;
    }

    public Coupon getCoupon() {
        return coupon;
    }

    public void use() {
        this.state = IssuedCouponState.USED;
    }

    public void revert() {
        this.state = IssuedCouponState.DOWNLOADED;
    }

    public void cancel() {
        this.state = IssuedCouponState.CANCELED;
    }

    public boolean isUsed() {
        return this.state == IssuedCouponState.USED;
    }

    public Principal getPrincipal() {
        return principal;
    }
}

File: core/core-domain/src/main/kotlin/com/coffee/gu/coupon/IssuedCoupon.kt
package com.coffee.gu.coupon

import com.coffee.gu.Principal
import com.coffee.gu.enums.IssuedCouponState

class IssuedCoupon(
    val id: Long = 0,
    val principal: Principal,
    var state: IssuedCouponState,
    val coupon: Coupon,
) {
    val isUsed: Boolean
        get() = state == IssuedCouponState.USED

    fun use() {
        this.state = IssuedCouponState.USED
    }

    fun revert() {
        this.state = IssuedCouponState.DOWNLOADED
    }

    fun cancel() {
        this.state = IssuedCouponState.CANCELED
    }

    companion object {
        @JvmStatic
        fun download(principal: Principal, coupon: Coupon): IssuedCoupon {
            return IssuedCoupon(
                id = 0,
                principal = principal,
                state = IssuedCouponState.DOWNLOADED,
                coupon = coupon
            )
        }
    }
}
